coordinator.go

```diff
@@ -2,69 +2,252 @@ package influxdb
+	ErrWriteFailed = errors.New("write failed")
+)
 
 // Coordinator handle queries and writes across multiple local and remote
 // data nodes.
 type Coordinator struct {
+	mu      sync.RWMutex
+	closing chan struct{}
+
+	MetaStore    meta.Store
+	shardWriters []ShardWriter
+}
+
+// ShardMapping contiains a mapping of a shards to a points.
+type ShardMapping struct {
+	Points map[uint64][]data.Point   // The points associated with a shard ID
+	Shards map[uint64]meta.ShardInfo // The shards that have been mapped, keyed by shard ID
+}
+
+func NewCoordinator() *Coordinator {
+	return &Coordinator{
+		closing: make(chan struct{}),
+	}
+}
+
+// NewShardMapping creates an empty ShardMapping
+func NewShardMapping() *ShardMapping {
+	return &ShardMapping{
+		Points: map[uint64][]data.Point{},
+		Shards: map[uint64]meta.ShardInfo{},
+	}
+}
+
+// MapPoint maps a point to shard
+func (s *ShardMapping) MapPoint(shardInfo meta.ShardInfo, p data.Point) {
+	points, ok := s.Points[shardInfo.ID]
+	if !ok {
+		s.Points[shardInfo.ID] = []data.Point{p}
+	} else {
+		s.Points[shardInfo.ID] = append(points, p)
+	}
+	s.Shards[shardInfo.ID] = shardInfo
+}
+
+// ShardWriter provides the ability to write a slice of points to s given shard ID.
+// It should return the number of times the set of points was written or an error
+// if the write failed.
+type ShardWriter interface {
+	WriteShard(shardID uint64, points []data.Point) (int, error)
+}
+
+func (c *Coordinator) Open() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.closing == nil {
+		c.closing = make(chan struct{})
+	}
+	return nil
+}
+
+func (c *Coordinator) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.closing != nil {
+		close(c.closing)
+		c.closing = nil
+	}
+	return nil
+}
+
+func (c *Coordinator) AddShardWriter(s ShardWriter) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.shardWriters = append(c.shardWriters, s)
+}
+
+// MapShards maps the points contained in wp to a ShardMapping.  If a point
+// maps to a shard group or shard that does not currently exist, it will be
+// created before returning the mapping.
+func (c *Coordinator) MapShards(wp *WritePointsRequest) (*ShardMapping, error) {
+
+	// holds the start time ranges for required shard groups
+	timeRanges := map[time.Time]*meta.ShardGroupInfo{}
+
+	rp, err := c.MetaStore.RetentionPolicy(wp.Database, wp.RetentionPolicy)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, p := range wp.Points {
+		timeRanges[p.Time.Truncate(rp.ShardGroupDuration)] = nil
+	}
+
+	// holds all the shard groups and shards that are required for writes
+	for t := range timeRanges {
+		g, err := c.MetaStore.CreateShardGroupIfNotExists(wp.Database, wp.RetentionPolicy, t)
+		if err != nil {
+			return nil, err
+		}
+		timeRanges[t] = g
+	}
+
+	shardMapping := NewShardMapping()
+	for _, p := range wp.Points {
+		g := timeRanges[p.Time.Truncate(rp.ShardGroupDuration)]
+		sid := p.SeriesID()
+		shardInfo := g.Shards[sid%uint64(len(g.Shards))]
```

Comment:
Should we extract this logic to a method `ShardGroupInfo` struct? We'll need to use it else where I think.